### PR TITLE
[BEAM-2033] Add HadoopResourceId

### PR DESCRIPTION
**Pull request by ssisk**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
merging person: I believe we'll want to do the implementation of this in a feature branch - this'd be a good time to create that feature branch. 

R: @tgroh 

It's worth noting that GcsResourceId spends a lot of time on worrying about slashes - that's not a thing we'll need for this since Hadoop's Path does a good job of cleaning it up, so all that's handled at a lower level.

### CODE REVIEW

**Comment by ssisk** [comment]
cc @jbonofre if you're interested in the review. I suggested opening a feature branch for HadoopFileSystem so it'd be easier to collaborate if you'd like to.

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java [hunk 1]
```diff
@@ -17,26 +17,68 @@
...
 
 /**
  * {@link ResourceId} implementation for the {@link HadoopFileSystem}.
  */
 public class HadoopResourceId implements ResourceId {
 
+  private final Path path;
+
+  /**
+   * You can only use absolute paths - if you have a relative path, you can use it with resolve() as
```
Comment (tgroh) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: "Constructs a HadoopResourceId from the provided absolute path. If only a relative path is available, (How to fix)"

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java [hunk 2]
```diff
@@ -17,26 +17,68 @@
...
+    this.path = path;
+  }
+
   @Override
   public ResourceId resolve(String other, ResolveOptions resolveOptions) {
-    throw new UnsupportedOperationException();
+    checkArgument(
+        resolveOptions.equals(ResolveOptions.StandardResolveOptions.RESOLVE_FILE)
+            || resolveOptions.equals(ResolveOptions.StandardResolveOptions.RESOLVE_DIRECTORY),
+        String.format("ResolveOptions: [%s] is not supported.", resolveOptions));
```
Comment (tgroh) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: What is supported?

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java [hunk 3]
```diff
@@ -17,26 +17,68 @@
...
+    if (resolveOptions.equals(ResolveOptions.StandardResolveOptions.RESOLVE_FILE)) {
+      checkArgument(
+          !other.endsWith("/"),
+          "The resolved file: [%s] should not end with '/'.", other);
```
Comment (tgroh) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: "The file to resolve"?

**Comment by jbonofre** [comment]
@ssisk  Thanks, I will !

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java (hunk 1 above)
Comment (ssisk) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: done

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java (hunk 2 above)
Comment (ssisk) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: done.

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java (hunk 3 above)
Comment (ssisk) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: I re-worded this.

**Comment by ssisk** [comment]
also - please merge into feature branch, not master

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java [hunk 4]
```diff
@@ -17,26 +17,71 @@
...
+
+  /**
+   * Constructs a HadoopResourceId from the provided absolute path. If only a relative path is
+   * available, you can create a {@link HadoopResourceId} from the absolute path of the root of the
+   * server, and then use resolve to add the relative path to the root.
+   * @param path
+   * @return
+   */
+  public static HadoopResourceId fromPath(Path path) {
+    checkNotNull(path);
```
Comment (tgroh) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: "path is required"

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java [hunk 5]
```diff
@@ -17,26 +17,71 @@
...
+    checkArgument(path.isAbsolute());
```
Comment (tgroh) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: path must be absolute

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java [hunk 6]
```diff
@@ -17,26 +17,71 @@
...
  */
 public class HadoopResourceId implements ResourceId {
 
+  private final Path path;
+
+  /**
+   * Constructs a HadoopResourceId from the provided absolute path. If only a relative path is
+   * available, you can create a {@link HadoopResourceId} from the absolute path of the root of the
+   * server, and then use resolve to add the relative path to the root.
+   * @param path
```
Comment (tgroh) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopResourceId.java]: rm

**Comment by ssisk** [comment]
fixed. PTAL

**Pull request closed without merging**